<a href="https://colab.research.google.com/github/fadhilatut/EDA-Internet-Price-in-200-Country-2022/blob/main/Final_Project_EDA_Fadhilatut.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### About
The dataset used for this EDA was from : https://www.kaggle.com/datasets/ramjasmaurya/1-gb-internet-price

Its a dataset of internet price aroun 200+ country ion 2022. it has 13 column with 243 rows of datas.

This EDA is my final project to complete the SADASA Academy scholarship - Data Science Basic with Python program.

The main purpose of this EDA is to explore about the dataset 

### Library

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.io as pio
from plotly.subplots import make_subplots
import plotly.graph_objects as go

### Import Dataset

Importing the dataset that i already put on my google drive. In this dataset, by default it already has its own index column, so i import it without the original column, to prevent double index.

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/dataset/all_csv sorted.csv',index_col=[0])
df_backup = df.copy()
df

Country code                           Country  Continental region  \
0             IL                            Israel           NEAR EAST   
1             KG                        Kyrgyzstan   CIS (FORMER USSR)   
2             FJ                              Fiji             OCEANIA   
3             IT                             Italy      WESTERN EUROPE   
4             SD                             Sudan  SUB-SAHARAN AFRICA   
..           ...                               ...                 ...   
239           VA           Vatican City (Holy See)              Europe   
240           VE                         Venezuela       SOUTH AMERICA   
241           WF                 Wallis and Futuna             OCEANIA   
242           CD  Democratic Republic of the Congo  SUB-SAHARAN AFRICA   
243           ZW                          Zimbabwe  SUB-SAHARAN AFRICA   

     NO. OF Internet Plans                 Average price of 1GB (USD)  \
0                      27.0                                      0.05   
1                      20.0                                      0.15   
2                      18.0                                      0.19   
3                      29.0                                      0.27   
4                      33.0                                      0.27   
..                      ...                                       ...   
239                     NaN                              NO PROVIDERS   
240                     NaN                            HYPERINFLATION   
241                     NaN                              NO PROVIDERS   
242                     NaN  Prices listed in non-convertible 'units'   
243                     NaN                 UNRELIABLE EXCHANGE RATES   

     Cheapest 1GB for 30 days (USD)  Most expensive 1GB (USD)  \
0                              0.02                     20.95   
1                              0.10                      7.08   
2                              0.05                      0.85   
3                              0.09                      3.54   
4                              0.03                      0.92   
..                              ...                       ...   
239                             NaN                       NaN   
240                             NaN                       NaN   
241                             NaN                       NaN   
242                             NaN                       NaN   
243                             NaN                       NaN   

    Average price of 1GB (USD  at the start of 2021)  \
0                                               0.11   
1                                               0.21   
2                                               0.59   
3                                               0.43   
4                                               0.63   
..                                               ...   
239                                              NaN   
240                                              NaN   
241                                              NaN   
242                                              NaN   
243                                              NaN   

    Average price of 1GB (USD – at start of 2020) Internet users  Population  \
0                                            0.90      6,788,737   8,381,516   
1                                            0.27      2,309,235   6,304,030   
2                                            3.57        452,479     883,483   
3                                            1.73     50,540,000  60,627,291   
4                                            0.68     12,512,639  41,801,533   
..                                            ...            ...         ...   
239                                           NaN            NaN         NaN   
240                                           NaN     20,564,451  28,887,118   
241                                           NaN          1,383

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 243 entries, 0 to 243
Data columns (total 12 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   Country code                                      243 non-null    object 
 1   Country                                           243 non-null    object 
 2   Continental region                                243 non-null    object 
 3   NO. OF Internet Plans                             231 non-null    float64
 4   Average price of 1GB (USD)                        243 non-null    object 
 5   Cheapest 1GB for 30 days (USD)                    231 non-null    float64
 6   Most expensive 1GB (USD)                          231 non-null    float64
 7   Average price of 1GB (USD  at the start of 2021)  231 non-null    object 
 8   Average price of 1GB (USD – at start of 2020)     231 non-null    object 
 9   Internet users       

From the information above, it show each column data type, some of them doesn't match for their values, ex: the column Average price of 1GB (USD) is supposed to only contain numeric values, but its datatype is object, it need to be explored and then converted to float64 datatype so it can be processed correctly.

### Preprocessing

First drop unnecessary column, i decided to drop these 2 column because i will only use the average price.

In [ ]:
df = df.drop(columns=['Cheapest 1GB for 30 days (USD)','Most expensive 1GB (USD)'])
df

Country code                           Country  Continental region  \
0             IL                            Israel           NEAR EAST   
1             KG                        Kyrgyzstan   CIS (FORMER USSR)   
2             FJ                              Fiji             OCEANIA   
3             IT                             Italy      WESTERN EUROPE   
4             SD                             Sudan  SUB-SAHARAN AFRICA   
..           ...                               ...                 ...   
239           VA           Vatican City (Holy See)              Europe   
240           VE                         Venezuela       SOUTH AMERICA   
241           WF                 Wallis and Futuna             OCEANIA   
242           CD  Democratic Republic of the Congo  SUB-SAHARAN AFRICA   
243           ZW                          Zimbabwe  SUB-SAHARAN AFRICA   

     NO. OF Internet Plans                 Average price of 1GB (USD)  \
0                      27.0                                      0.05   
1                      20.0                                      0.15   
2                      18.0                                      0.19   
3                      29.0                                      0.27   
4                      33.0                                      0.27   
..                      ...                                       ...   
239                     NaN                              NO PROVIDERS   
240                     NaN                            HYPERINFLATION   
241                     NaN                              NO PROVIDERS   
242                     NaN  Prices listed in non-convertible 'units'   
243                     NaN                 UNRELIABLE EXCHANGE RATES   

    Average price of 1GB (USD  at the start of 2021)  \
0                                               0.11   
1                                               0.21   
2                                               0.59   
3                                               0.43   
4                                               0.63   
..                                               ...   
239                                              NaN   
240                                              NaN   
241                                              NaN   
242                                              NaN   
243                                              NaN   

    Average price of 1GB (USD – at start of 2020) Internet users  Population  \
0                                            0.90      6,788,737   8,381,516   
1                                            0.27      2,309,235   6,304,030   
2                                            3.57        452,479     883,483   
3                                            1.73     50,540,000  60,627,291   
4                                            0.68     12,512,639  41,801,533   
..                                            ...            ...         ...   
239                                           NaN            NaN         NaN   
240                                           NaN     20,564,451  28,887,118   
241                                           NaN          1,383      11,661   
242                                           NaN      7,011,507  84,068,091   
243                                           NaN      4,472,992  14,438,802   

     Avg \n(Mbit/s)Ookla  
0                  28.01  
1                  16.30  
2                  25.99  
3                  37.15  
4                   9.50  
..                   ...  
239                  NaN  
240                 5.68  
241                  NaN  
242                12.08  
243                13.99  

[243 rows x 10 columns]

Then i checked the NaN values for each column

In [ ]:
df.isnull().sum().sort_values(ascending=False)

Avg \n(Mbit/s)Ookla                                 102
Population                                           32
Internet users                                       30
NO. OF Internet Plans                                12
Average price of 1GB (USD  at the start of 2021)     12
Average price of 1GB (USD – at start of 2020)        12
Country code                                          0
Country                                               0
Continental region                                    0
Average price of 1GB (USD)                            0
dtype: int64

3 of 12 column has same NaN value, and after look the datas the NaN values were on the same row, so i drop the 12 row, because those rows is contain unavaiable values

In [ ]:
df = df[pd.notnull(df['NO. OF Internet Plans '])]
df.isnull().sum().sort_values(ascending=False)

Avg \n(Mbit/s)Ookla                                 93
Population                                          28
Internet users                                      26
Country code                                         0
Country                                              0
Continental region                                   0
NO. OF Internet Plans                                0
Average price of 1GB (USD)                           0
Average price of 1GB (USD  at the start of 2021)     0
Average price of 1GB (USD – at start of 2020)        0
dtype: int64

# Change Data Type

After i checked the datas in each column,all row in these 2 column were contain numeric values, yet it still labelled as object, because the numeric values contain comma (,) and it counted as string. Therefore i remove all comas in these 2 column

In [ ]:
df["Population"] = [float(str(i).replace(",", "")) for i in df["Population"]]
df["Internet users"] = [float(str(i).replace(",", "")) for i in df["Internet users"]]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In these 2 columns, i found out they had special values in both of them. it was the "NO PACKAGES" values. then i replace those value with 0, because No Pacakages can also mean 0 

In [ ]:
df['Average price of 1GB (USD  at the start of 2021)'] = df['Average price of 1GB (USD  at the start of 2021)'].replace(['NO PACKAGES'],0)
df['Average price of 1GB (USD – at start of 2020)'] = df['Average price of 1GB (USD – at start of 2020)'].replace(['NO PACKAGES'],0)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


After replace the value in those columns to all numeric, i change the column datatype to float64

In [ ]:
df['Average price of 1GB (USD  at the start of 2021)']=df['Average price of 1GB (USD  at the start of 2021)'].astype('float')
df['Average price of 1GB (USD – at start of 2020)'] = df['Average price of 1GB (USD – at start of 2020)'].astype('float')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
df['Average price of 1GB (USD)']= df['Average price of 1GB (USD)'].astype('float')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 231 entries, 0 to 232
Data columns (total 10 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   Country code                                      231 non-null    object 
 1   Country                                           231 non-null    object 
 2   Continental region                                231 non-null    object 
 3   NO. OF Internet Plans                             231 non-null    float64
 4   Average price of 1GB (USD)                        231 non-null    float64
 5   Average price of 1GB (USD  at the start of 2021)  231 non-null    float64
 6   Average price of 1GB (USD – at start of 2020)     231 non-null    float64
 7   Internet users                                    205 non-null    float64
 8   Population                                        203 non-null    float64
 9   Avg 
(Mbit/s)Ookla   

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


After all column datatypes were changed correctly, i check duplicate data in some column. i found duplicate in column Country, and then drop the duplicate

In [ ]:
df.shape

(231, 10)

In [ ]:
df.duplicated(subset='Country').sum()

2

In [ ]:
df = df.drop_duplicates(subset='Country')

Check the duplicated values

In [ ]:
df.duplicated(subset='Country').sum()

0

Resetting the index after drop the duplicate value

In [ ]:
df = df.reset_index(drop=True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 229 entries, 0 to 228
Data columns (total 10 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   Country code                                      229 non-null    object 
 1   Country                                           229 non-null    object 
 2   Continental region                                229 non-null    object 
 3   NO. OF Internet Plans                             229 non-null    float64
 4   Average price of 1GB (USD)                        229 non-null    float64
 5   Average price of 1GB (USD  at the start of 2021)  229 non-null    float64
 6   Average price of 1GB (USD – at start of 2020)     229 non-null    float64
 7   Internet users                                    203 non-null    float64
 8   Population                                        201 non-null    float64
 9   Avg 
(Mbit/s)Ookla   

View the outlier data. the last step of this preprocessing is filling the remaining NaN value, before that i check the data outliers

In [ ]:
vars = ['Avg \n(Mbit/s)Ookla','Internet users','Population']
fig = make_subplots(rows=1, cols=len(vars))
for i, var in enumerate(vars):
  fig.add_trace(
      go.Box(y=df[var],name=var),
      row=1,col=i+1
  )
fig.update_traces()


Replacing NaN value with median of each column. 

In [ ]:
#Replcae NaN value
nan_replace = {'Internet users':df['Internet users'].median(),'Population':df['Population'].median(),\
               'Avg \n(Mbit/s)Ookla':df['Avg \n(Mbit/s)Ookla'].median()}
df = df.fillna(nan_replace)

Check remaining NaN values

In [ ]:
df.isna().sum()

Country code                                        0
Country                                             0
Continental region                                  0
NO. OF Internet Plans                               0
Average price of 1GB (USD)                          0
Average price of 1GB (USD  at the start of 2021)    0
Average price of 1GB (USD – at start of 2020)       0
Internet users                                      0
Population                                          0
Avg \n(Mbit/s)Ookla                                 0
dtype: int64

In [ ]:
df

Country code                                       Country  \
0             IL                                        Israel   
1             KG                                    Kyrgyzstan   
2             FJ                                          Fiji   
3             IT                                         Italy   
4             SD                                         Sudan   
..           ...                                           ...   
224           MW                                        Malawi   
225           ST                         São Tomé and Príncipe   
226           SH  Saint Helena, Ascension and Tristan da Cunha   
227           FK                              Falkland Islands   
228           GQ                             Equatorial Guinea   

     Continental region  NO. OF Internet Plans   Average price of 1GB (USD)  \
0             NEAR EAST                    27.0                        0.05   
1     CIS (FORMER USSR)                    20.0                        0.15   
2               OCEANIA                    18.0                        0.19   
3        WESTERN EUROPE                    29.0                        0.27   
4    SUB-SAHARAN AFRICA                    33.0                        0.27   
..                  ...                     ...                         ...   
224  SUB-SAHARAN AFRICA                     5.0                       25.46   
225  SUB-SAHARAN AFRICA                    16.0                       30.97   
226  SUB-SAHARAN AFRICA                     4.0                       39.87   
227       SOUTH AMERICA                     4.0                       44.56   
228  SUB-SAHARAN AFRICA                     4.0                       49.67   

     Average price of 1GB (USD  at the start of 2021)  \
0                                                0.11   
1                                                0.21   
2                                                0.59   
3                                                0.43   
4                                                0.63   
..                                                ...   
224                                             27.41   
225                                             28.26   
226                                             52.50   
227                                             40.41   
228                                             12.78   

     Average price of 1GB (USD – at start of 2020)  Internet users  \
0                                             0.90       6788737.0   
1                                             0.27       2309235.0   
2                                             3.57        452479.0   
3                                             1.73      50540000.0   
4                                             0.68      12512639.0   
..                                             ...             ...   
224                                           3.59       2566126.0   
225                                           5.33         61155.0   
226                                           6.04          2906.0   
227                                          55.47          2881.0   
228                                          47.39        332642.0   

     Population  Avg \n(Mbit/s)Ookla  
0     8381516.0                28.01  
1     6304030.0                16.30  
2      883483.0                25.99  
3    60627291.0                37.15  
4    41801533.0                 9.50  
..          ...                  ...  
224  18143217.0                17.95  
225    211028.0                28.19  
226      6035.0                28.19  
227      3234.0                28.19  
228   1308975.0                28.19  

[229 rows x 10 columns]

After data is cleaned, it have 229 row and 13 column

### EDA

Question:
1. How is the illustration of internet speed around the world? 
2. What country has the highest & lowest average internet speed?
3. What country has the highest & cheapest average internet price per 1 GB?
4. How is the percentage of internet user around the world?
5. How is the average price changes in 2020 - 2021? does most country have price increase or decrease?
6. How much is the average price differences in 2020-2021?
7. How the correlation of each column?


# 1. The Illustration of internet speed around the world

In [ ]:
fig = px.choropleth(df, locations='Country', locationmode='country names',scope='world',color='Avg \n(Mbit/s)Ookla', color_continuous_scale= 'sunsetdark')
fig.update_layout(margin={'r':0,'t':0,'l':0,'b':0}, coloraxis_colorbar=dict(
    title = 'Avg Speed (Mbit/s)',
    ticks = 'outside',
    tickvals = [1,5,10,20,30,40,50,60,70,80,90,100,125,135,140],
    dtick = 7
))
fig.show()

From image above, there are domination of high internet speed in north america, australia, part of europe, and part of asia

# 2. Country with highest and lowest Internet Speed

In [ ]:
df_spd = df[df['Avg \n(Mbit/s)Ookla'] == df['Avg \n(Mbit/s)Ookla'].max()]
df_spd2 = df[df['Avg \n(Mbit/s)Ookla'] == df['Avg \n(Mbit/s)Ookla'].min()]
print ("Country With Highest Internet Speed")
df_spd[['Country','Avg \n(Mbit/s)Ookla']]

Country With Highest Internet Speed


Country  Avg \n(Mbit/s)Ookla
204  United Arab Emirates               135.35

In [ ]:
print ("Country With slowest Internet Speed")
df_spd2[['Country','Avg \n(Mbit/s)Ookla']]

Country With slowest Internet Speed


Country  Avg \n(Mbit/s)Ookla
92  Palestine                 4.89

This is the top 10 of Highest & Lowest Internet Speed around the world

In [ ]:
df_spd = df.sort_values('Avg \n(Mbit/s)Ookla',ascending=False)
negara = df_spd['Country']
negara = negara.iloc[:10]
kecepatan = df_spd['Avg \n(Mbit/s)Ookla'].iloc[:10]
px.bar(x=negara,y=kecepatan)

In [ ]:
df_spd2 = df.sort_values('Avg \n(Mbit/s)Ookla',ascending=True)
negara = df_spd2['Country']
negara = negara.iloc[:10]
kecepatan = df_spd2['Avg \n(Mbit/s)Ookla'].iloc[:10]
px.bar(x=negara,y=kecepatan)

# 3. Country with highest & lowest internet price

In [ ]:
df_price =df[df['Average price of 1GB (USD)'] == df['Average price of 1GB (USD)'].max()]
df_price2 =df[df['Average price of 1GB (USD)'] == df['Average price of 1GB (USD)'].min()]
print ("Country with highest Average Internet Price per 1GB")
df_price[['Country','Average price of 1GB (USD)']]

Country with highest Average Internet Price per 1GB


Country  Average price of 1GB (USD)
228  Equatorial Guinea                       49.67

In [ ]:
print ("Country with lowest Average Internet Price per 1 GB")
df_price2[['Country','Average price of 1GB (USD)']]

Country with lowest Average Internet Price per 1 GB


Country  Average price of 1GB (USD)
0  Israel                        0.05

# 4. Percentage of Internet User

In [ ]:
dfbackup1 = df.copy()
dfbackup1

Country code                                       Country  \
0             IL                                        Israel   
1             KG                                    Kyrgyzstan   
2             FJ                                          Fiji   
3             IT                                         Italy   
4             SD                                         Sudan   
..           ...                                           ...   
224           MW                                        Malawi   
225           ST                         São Tomé and Príncipe   
226           SH  Saint Helena, Ascension and Tristan da Cunha   
227           FK                              Falkland Islands   
228           GQ                             Equatorial Guinea   

     Continental region  NO. OF Internet Plans   Average price of 1GB (USD)  \
0             NEAR EAST                    27.0                        0.05   
1     CIS (FORMER USSR)                    20.0                        0.15   
2               OCEANIA                    18.0                        0.19   
3        WESTERN EUROPE                    29.0                        0.27   
4    SUB-SAHARAN AFRICA                    33.0                        0.27   
..                  ...                     ...                         ...   
224  SUB-SAHARAN AFRICA                     5.0                       25.46   
225  SUB-SAHARAN AFRICA                    16.0                       30.97   
226  SUB-SAHARAN AFRICA                     4.0                       39.87   
227       SOUTH AMERICA                     4.0                       44.56   
228  SUB-SAHARAN AFRICA                     4.0                       49.67   

     Average price of 1GB (USD  at the start of 2021)  \
0                                                0.11   
1                                                0.21   
2                                                0.59   
3                                                0.43   
4                                                0.63   
..                                                ...   
224                                             27.41   
225                                             28.26   
226                                             52.50   
227                                             40.41   
228                                             12.78   

     Average price of 1GB (USD – at start of 2020)  Internet users  \
0                                             0.90       6788737.0   
1                                             0.27       2309235.0   
2                                             3.57        452479.0   
3                                             1.73      50540000.0   
4                                             0.68      12512639.0   
..                                             ...             ...   
224                                           3.59       2566126.0   
225                                           5.33         61155.0   
226                                           6.04          2906.0   
227                                          55.47          2881.0   
228                                          47.39        332642.0   

     Population  Avg \n(Mbit/s)Ookla  
0     8381516.0                28.01  
1     6304030.0                16.30  
2      883483.0                25.99  
3    60627291.0                37.15  
4    41801533.0                 9.50  
..          ...                  ...  
224  18143217.0                17.95  
225    211028.0                28.19  
226      6035.0                28.19  
227      3234.0                28.19  
228   1308975.0                28.19  

[229 rows x 10 columns]

In [ ]:

df_peruser = df
df_peruser['Percentage Internet User'] = (df_peruser['Internet users']/df_peruser['Population']) * 100
df_peruser1 = df_peruser[df_peruser['Percentage Internet User']== df_peruser['Percentage Internet User'].max()]
df_peruser2 = df_peruser[df_peruser['Percentage Internet User']== df_peruser['Percentage Internet User'].min()]
df_peruser1[['Country','Internet users','Population','Percentage Internet User']]




Country  Internet users  Population  Percentage Internet User
133  Faroe Islands         48097.0     48497.0                 99.175207

In [ ]:
df_peruser2[['Country','Internet users','Population','Percentage Internet User']]

Country  Internet users  Population  Percentage Internet User
72  Jersey         38958.0   7051608.0                   0.55247

In [ ]:
px.bar(df_peruser, y ='Percentage Internet User',x='Country', title='Percentage of Intenet User Chart')

# 5. Price changes in 2020-2021

In [ ]:
price = df['Average price of 1GB (USD  at the start of 2021)']-df['Average price of 1GB (USD – at start of 2020)']
df['Price_diff'] = price
df.loc[df['Price_diff']< 0, 'S_Price'] = 'Decrease'
df.loc[df['Price_diff']> 0, 'S_Price'] = 'Increase'
df.loc[df['Price_diff']== 0, 'S_Price'] = 'Fixed'
dfa = df['S_Price'].value_counts()
#px.bar(dfa)
dfa

Decrease    185
Increase     43
Fixed         1
Name: S_Price, dtype: int64

There are 185 country that the average price of internet were decreased in 2020-2021, 43 country that the average price of internet were increased, and 1 country with fixed price

#6. Average price differences in 2020-2021

In [ ]:
prc_plus = (df['Price_diff'] / df['Average price of 1GB (USD – at start of 2020)'])*100
prc_min = (df['Price_diff'] / df['Average price of 1GB (USD  at the start of 2021)'])*100
prc_min

df.loc[df['Price_diff']< 0, 'Per_Price_diff (%)'] = prc_plus
df.loc[df['Price_diff']> 0, 'Per_Price_diff (%)'] = prc_min
df.loc[df['Price_diff']== 0, 'Per_Price_diff (%)'] = 0

In [ ]:
df.tail(10)

Country code                                       Country  \
219           YE                                         Yemen   
220           BM                                       Bermuda   
221           TK                                       Tokelau   
222           TM                                  Turkmenistan   
223           TD                                          Chad   
224           MW                                        Malawi   
225           ST                         São Tomé and Príncipe   
226           SH  Saint Helena, Ascension and Tristan da Cunha   
227           FK                              Falkland Islands   
228           GQ                             Equatorial Guinea   

     Continental region  NO. OF Internet Plans   Average price of 1GB (USD)  \
219           NEAR EAST                    17.0                       15.98   
220    NORTHERN AMERICA                    21.0                       19.80   
221             OCEANIA                     6.0                       20.48   
222   CIS (FORMER USSR)                     4.0                       21.41   
223  SUB-SAHARAN AFRICA                    15.0                       23.33   
224  SUB-SAHARAN AFRICA                     5.0                       25.46   
225  SUB-SAHARAN AFRICA                    16.0                       30.97   
226  SUB-SAHARAN AFRICA                     4.0                       39.87   
227       SOUTH AMERICA                     4.0                       44.56   
228  SUB-SAHARAN AFRICA                     4.0                       49.67   

     Average price of 1GB (USD  at the start of 2021)  \
219                                             15.98   
220                                             28.75   
221                                             17.75   
222                                             11.44   
223                                             23.33   
224                                             27.41   
225                                             28.26   
226                                             52.50   
227                                             40.41   
228                                             12.78   

     Average price of 1GB (USD – at start of 2020)  Internet users  \
219                                          15.73       7548512.0   
220                                          37.74         60349.0   
221                                          29.96       2664928.0   
222                                          19.81       1223591.0   
223                                          23.33        968500.0   
224                                           3.59       2566126.0   
225                                           5.33         61155.0   
226                                           6.04          2906.0   
227                                          55.47          2881.0   
228                                          47.39        332642.0   

     Population  Avg \n(Mbit/s)Ookla  Percentage Internet User  Price_diff  \
219  28498683.0                16.89                 26.487231        0.25   
220     62756.0                28.19                 96.164510       -8.99   
221   7051608.0                28.19                 37.791777      -12.21   
222   5850901.0                28.19                 20.912865       -8.37   
223  15477729.0                28.19                  6.257378        0.00   
224  18143217.0                17.95                 14.143721       23.82   
225    211028.0                28.19                 28.979567       22.93   
226      6035.0                28.19                 48.152444       46.46   
227      3234.0                28.19                 89.084725      -15.06   
228   1308975.0                28.19                 25.412403      -34.61   

      S_Price  Per_Price_diff (%)  
219  Increase            1.564456  
220  Decrease          -23.820880  
221  Decrease          -40.754339  
222  

In [ ]:
px.bar(df, y ='Per_Price_diff (%)',x='Country', title='Percentage of Price changes 2020-2021')